lib/oli/authoring/editing/container_editor.ex

```diff
@@ -264,6 +266,102 @@ defmodule Oli.Authoring.Editing.ContainerEditor do
     end
   end
 
+  @doc """
+    Duplicates a page within its project.
+    A true deep-copy of all the content is made. Resources and Revisions are created instead of simply referenced.
+  """
+  def duplicate_page(
+        %Revision{} = container,
+        page_id,
+        %Author{} = author,
+        %Project{} = project
+      ) do
+    original_page =
+      Resources.get_revision!(page_id)
+      |> Map.from_struct()
+
+    new_page_attrs =
+      original_page
+      |> Map.delete(:slug)
+      |> Map.delete(:inserted_at)
+      |> Map.delete(:updated_at)
+      |> Map.delete(:updated_at)
+      |> Map.delete(:resource_id)
+      |> Map.delete(:resource)
+      |> Map.put(:title, "Copy of #{original_page.title}")
+      |> Map.put(:content, nil)
+
+    Repo.transaction(fn ->
+      with {:ok, created_revision} <- add_new(container, new_page_attrs, author, project),
+           {:ok, duplicated_content} <-
+             duplicate_content_activities(
+               original_page.content,
+               project.slug,
+               author
+             ),
+           {:ok, updated_revision} <-
+             Resources.update_revision(created_revision, %{content: duplicated_content}) do
+        updated_revision
+      else
+        {:error, e} -> Repo.rollback(e)
+      end
+    end)
+  end
+
+  defp duplicate_content_activities(content, project_slug, author) do
+    case deep_copy_activities(content["model"], project_slug, author) do
+      {:ok, duplicated_activities} ->
+        {:ok, %{content | "model" => Enum.reverse(duplicated_activities)}}
+
+      {:error, error} ->
+        {:error, error}
+    end
+  end
+
+  defp deep_copy_activities([], _project_slug, _author), do: {:ok, []}
+
+  defp deep_copy_activities(model, project_slug, author) do
+    Enum.reduce_while(model, {:ok, []}, fn activity, {:ok, acc} ->
+      case deep_copy_activity(activity, project_slug, author) do
+        {:ok, new_activity} -> {:cont, {:ok, [new_activity | acc]}}
+        {:error, _} = err -> {:halt, err}
+      end
+    end)
+  end
+
+  defp deep_copy_activity(%{"type" => "activity-reference"} = item, project_slug, author) do
+    activity_revision = AuthoringResolver.from_resource_id(project_slug, item["activity_id"])
+
+    activity_type = Activities.get_registration(activity_revision.activity_type_id)
+
+    objectives =
+      Enum.reduce(activity_revision.objectives, [], fn {_key, ids}, acc -> acc ++ ids end)
```

Comment:
Is the reduce necessary or we just link the same objectives that are in `activity_revision`?

Reply:
In certain cases, this approach for handling the objectives is not going to lead to correct duplication, given that the `activity_editor.create/7` implementation takes the list of supplied objectives and applies them to *all* parts of the activity.  In the original activity, each part may have had a different objective.  For example, consider the following `objectives` value for the activity that is being duplicated:

```
%{
  "1" => [100, 101, 102],
 "2" => [200, 201, 202],
}
```

In the above, we see this activity has two parts, identified by `"1"` and `"2"` respectively.  Each part has three objectives attached to them.  As-is, the current duplication code would reduce this structure down to a list of `[100, 101, 102, 200, 201, 202]` and the `attach_objectives_to_all_parts/2` impl within `activity_editor` would create the following:

```
%{
  "1" => [100, 101, 102, 200, 201, 202],
 "2" => [100, 101, 102, 200, 201, 202]
}
```

Reply:
@darrensiegel to further understand this (trying to cover all scenarios);
Am I right to say that non-adaptive activities will only have one part in the content? Didn't manage to replicate a scenario where I could assign an objective to a specific part for normal pages. It also seems like `ActivityEditor.create` will always receive an empty objectives list for non-adaptive lessons. Since the objectives are attached on an update and not when first adding the activity to the page. Perhaps that's not needed when creating from the UI?